stream_alert/shared/rule.py

```diff
@@ -0,0 +1,187 @@
+        self.outputs = kwargs.get('outputs')
+        self.matchers = kwargs.get('matchers')
+        self.datatypes = kwargs.get('datatypes')
+        self.req_subkeys = kwargs.get('req_subkeys')
+        self.initial_context = kwargs.get('context')
+        self.context = None
+        self.disabled = False
+
+        if not (self.logs or self.datatypes):
+            raise RuleInvalid(
+                "Invalid rule [{}] - rule must have either 'logs' "
+                "or 'datatypes' declared'".format(self.rule_name)
+            )
+
+        if self.rule_name in Rule._rules:
+            raise RuleInvalid('Rule [{}] already defined'.format(self.rule_name))
+
+        Rule._rules[self.rule_name] = self
+
+    def __str__(self):
+        return '<Rule: {}; outputs: {}; disabled: {}>'.format(
+            self.rule_name,
+            self.outputs,
+            self.disabled
+        )
+
+    def __repr__(self):
+        return self.__str__()
+
+    def check_matchers(self, record):
+        """Run any rule matchers against the record
+
+        Args:
+            record (dict): Record that this rule should be run against
+
+        Returns:
+            bools: True if all matchers apply to this record, False otherwise
+        """
+        if not self.matchers:
+            return True
+
+        return all(Matcher.process(matcher_name, record) for matcher_name in self.matchers)
+
+    @time_rule
+    def process(self, record):
+        """Process will call this rule's function on the passed record
+
+        Args:
+            record (dict): Record that this rule should be run against
+
+        Returns:
+            bool: True if this rule triggers for the passed record, False otherwise
+        """
+        try:
+            # The initial_context object must be copied. This avoids
+            # bleed over from other runs of the rule using the same
+            # context object
+            if self.initial_context is not None:
+                self.context = self.initial_context.copy()
+                return self.func(record, self.context)
+
+            return self.func(record)
+        except Exception:  # pylint: disable=broad-except
+            LOGGER.exception('Encountered error with rule: %s', self.rule_name)
+
+        return False
+
+    @classmethod
+    def disable(cls, name):
+        cls._rules[name].disabled = True
+
+    @classmethod
+    def rule_names(cls):
+        return Rule._rules.keys()
+
+    @classmethod
+    def get_rule(cls, rule_name):
+        return Rule._rules.get(rule_name)
+
+    @classmethod
+    def rules_with_datatypes(cls):
+        return [item for item in Rule._rules.values()
+                if item.datatypes and not item.disabled]
+
+    @classmethod
+    def rules_for_log_type(cls, log_type):
+        return [item for item in Rule._rules.values()
+                if (item.logs is None or log_type in item.logs) and not item.disabled]
+
+
+class MatcherInvalid(Exception):
```

Comment:
As above, my personal inclination here is `InvalidMatcherError`, but I don't feel strongly either way